feat: Login with mobile number with optional OTP Config

bhushan-barbuddhe opened the pull request:
# 📋 Overview

This PR introduces **passwordless mobile OTP authentication**, enabling users to log in using only their mobile number and SMS OTP verification. This authentication method enhances accessibility and offers a seamless alternative to traditional password-based login.

---

# ✨ Features Added

### 🔐 Passwordless Mobile OTP Authentication
- Login using mobile number and SMS OTP without requiring a password

### 🎨 Enhanced Login UI
- Separate login forms for email/password and mobile OTP authentication
- Dynamic form rendering based on system configuration settings
- Clean, responsive design optimized for mobile devices

---

# 🔄 Authentication Flow

**Mobile OTP Login Process:**
1. User enters their mobile number
2. System locates user by `mobile_no`
3. Generate a TOTP token and cache it with a temporary ID (`tmp_id`)
4. Send OTP via the existing SMS infrastructure
5. User enters the received OTP
6. Verify OTP using `confirm_otp_token()`
7. On successful verification, user is logged in and redirected to the app

---

# 🔧 Configuration Requirements

**System Settings:**
- `allow_login_using_mobile_number = 1`
- `allow_mobile_login_with_otp = 1`
- SMS settings properly configured for OTP delivery

**Optional Settings:**
- `disable_user_pass_login = 1` (enables mobile OTP-only login mode)

---

# 🛡️ Security Features

- Rate limiting: max 5 attempts per 10 minutes per mobile number
- Login attempt tracking with user locking on repeated failures
- Token expiration set to 300 seconds (5 minutes)
- Encrypted storage of OTP secrets
- HOTP verification using cached tokens for secure OTP validation

<img width="300" height="300" alt="Screenshot From 2025-08-09 17-07-03" src="https://github.com/user-attachments/assets/20c2e961-7379-46ab-a588-9fedffb14c94" />
<img width="300" height="300" alt="Screenshot From 2025-08-09 17-07-35" src="https://github.com/user-attachments/assets/de37739c-fdc5-4800-8ec7-86df3a97ec3b" />
<img width="300" height="300" alt="Screenshot From 2025-08-09 17-07-46" src="https://github.com/user-attachments/assets/568d2388-6ca3-453a-b631-e316158e84ee" />
<img width="300" height="300" alt="Screenshot From 2025-08-12 17-27-26" src="https://github.com/user-attachments/assets/2520de1f-83dc-44c6-bd0e-f505ec09e827" />
<img width="300" height="300" alt="Screenshot From 2025-08-12 17-28-21" src="https://github.com/user-attachments/assets/74e8c5be-a7b9-4403-a5d0-d46928f8f525" />


closes #33256 

https://docs.frappe.io/framework/user/en/api/utils?editWiki=1&wikiPagePatch=db3ts9c9e7

bhushan-barbuddhe committed:
feat: allow login using mobile number in system setting

bhushan-barbuddhe committed:
WIP:Login with OTP

bhushan-barbuddhe committed:
feat: Login with mobile number funtionlity

bhushan-barbuddhe committed:
fix: rate limit added

bhushan-barbuddhe committed:
Merge pull request #1 from dhwani-ris/feat/passwordless-mobile-login

Feat/passwordless mobile login

bhushan-barbuddhe committed:
minor

bhushan-barbuddhe committed:
Merge pull request #2 from dhwani-ris/feat/passwordless-mobile-login

minor

bhushan-barbuddhe committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe committed:
feat: add hook for custom SMS OTP sender (mobile_otp_sms_sender) in 2FA

bhushan-barbuddhe committed:
Merge branch 'feat/hook_send_token_sms' of https://github.com/dhwani-ris/frappe into develop

bhushan-barbuddhe committed:
feat: hook implementation for for send sms OTP

bhushan-barbuddhe committed:
Merge pull request #3 from dhwani-ris/feat/passwordless-mobile-login

feat: hook implementation for for send sms OTP

bhushan-barbuddhe committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe committed:
[ADD]: Hooks for send SMS

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+
+
+def validate_mobile_otp_prerequisites():
+	"""Validate that all prerequisites for mobile OTP login are met."""
+	if not cint(frappe.get_system_settings("allow_login_using_mobile_number")):
+		frappe.throw(_("Mobile login is not enabled."), frappe.AuthenticationError)
+	
+	if not cint(frappe.get_system_settings("allow_mobile_login_with_otp")):
+		frappe.throw(_("Mobile OTP login is not enabled."), frappe.AuthenticationError)
```
Can we reuse the above function?

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+		frappe.throw(_("Mobile OTP login is not enabled."), frappe.AuthenticationError)
+	
+	# Check SMS settings using existing validation
+	sms_gateway_url = frappe.db.get_single_value("SMS Settings", "sms_gateway_url")
```
Let's do `get_cached_value`.

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+		)
+
+
+def find_user_by_mobile(mobile_no):
```
```suggestion
def find_user_by_mobile(mobile_no: str) -> Dict:
```

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+	return user
+
+
+def generate_mobile_otp(user):
```
```suggestion
def generate_mobile_otp(user: str):
```

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+
+def generate_mobile_otp(user):
+	"""Generate OTP for mobile login using existing secure methods."""
+	# Use existing OTP secret generation (encrypted storage)
```
Unnecessary comment

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+	# Use existing OTP secret generation (encrypted storage)
+	otp_secret = get_otpsecret_for_(user)
+	
+	# Generate token using existing method
```
Not required.

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+	return token, otp_secret
+
+
+def cache_mobile_otp_data(user, token, otp_secret, tmp_id):
```
Add type annotations.

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+
+def cache_mobile_otp_data(user, token, otp_secret, tmp_id):
+	"""Cache mobile OTP data without requiring password."""
+	verification_method = get_verification_method()
+
+	pipeline = frappe.cache.pipeline()
+
+	# Set increased expiry time for SMS
+	if verification_method == "SMS":
+		expiry_time = frappe.flags.token_expiry or 300
+		pipeline.set(tmp_id + "_token", token, expiry_time)
+	else:
+		expiry_time = frappe.flags.otp_expiry or 180
+	
+	# Ensure all values are strings (not bytes)
+	user = str(user) if user else ""
+	otp_secret = str(otp_secret) if otp_secret else ""
+	
+	# Cache user and OTP secret without password
+	for k, v in {"_usr": user, "_otp_secret": otp_secret}.items():
+		pipeline.set(f"{tmp_id}{k}", v, expiry_time)
+	pipeline.execute()
```
Can we refactor 2FA OTP code to reuse here?

bhushan-barbuddhe replied:
Improved this function while ensuring it continues to operate independently without requiring 2FA to be enabled.

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+
+
+def send_mobile_login_otp(user, mobile_no):
+	"""Send OTP for mobile login using existing SMS infrastructure."""
+	validate_mobile_otp_prerequisites()
+	
+	# Check rate limiting using existing tracker
+	user_tracker = get_login_attempt_tracker(user, raise_locked_exception=True)
+	
+	# Generate OTP using existing secure method
+	token, otp_secret = generate_mobile_otp(user)
+	
+	# Generate session ID
+	tmp_id = frappe.generate_hash(length=8)
+	
+	# Cache OTP data using mobile-specific method (no password required)
+	cache_mobile_otp_data(user, token, otp_secret, tmp_id)
+	
+	# Hook support for custom SMS sender
+	hook_methods = frappe.get_hooks("mobile_otp_sms_sender")
+	if hook_methods:
+		status = frappe.get_attr(hook_methods[-1])(otp_secret, token=token, phone_no=mobile_no)
+	else:
+		status = send_token_via_sms(otp_secret, token=token, phone_no=mobile_no)
+
+	if not status:
+		frappe.throw(_("Failed to send OTP. Please try again."), frappe.AuthenticationError)
+	
+	# Format mobile number for display (security)
+	masked_mobile = mobile_no[:4] + "******" + mobile_no[-3:] if len(mobile_no) > 7 else "******"
```
Let's refactor 2FA to make it reusable.

bhushan-barbuddhe replied:
Improved this function while ensuring it continues to operate independently without requiring 2FA to be enabled.

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,178 @@
+		frappe.throw(_("Invalid OTP. Please try again."), frappe.AuthenticationError)
+
+
+def cleanup_expired_mobile_otp_sessions():
+	"""Cleanup expired mobile OTP sessions - handled by cache TTL automatically."""
+	# No need for manual cleanup as cache entries have TTL
+	pass 
```
To be removed.

NagariaHussain commented on frappe/www/login.html:
```diff
@@ -127,6 +127,16 @@ <h4>{{ _("Javascript is disabled on your browser") }}</h4>
+							<div class="login-button-wrapper">
+								<a href="#login-with-mobile-otp-link"
+									class="btn btn-block btn-default btn-sm btn-login-option btn-login-with-mobile-otp-link">
+									{{ _("login with Mobile OTP") }}</a>
```
```suggestion
									{{ _("Login with SMS OTP") }}</a>
```

NagariaHussain commented on frappe/www/login.html:
```diff
@@ -235,6 +246,50 @@ <h4>{{ _("Javascript is disabled on your browser") }}</h4>
+
+	<!-- Mobile OTP Login Section -->
+	<section class='for-login-with-mobile-otp-link'>
+		{{ logo_section(_('Login with Mobile OTP')) }}
```
```suggestion
		{{ logo_section(_('Login with SMS OTP')) }}
```

NagariaHussain commented on frappe/www/login.py:
```diff
@@ -118,6 +118,9 @@ def get_context(context):
 		else None
 	)
 
+	# Mobile OTP login settings
```
Comment not needed here too.

NagariaHussain commented on frappe/www/login.py:
```diff
@@ -216,3 +219,75 @@ def sanitize_redirect(redirect: str | None) -> str | None:
+	frappe.local.login_manager.login_as(user)
+	
+	# Set response in the format expected by frontend
+	desk_user = frappe.db.get_value("User", frappe.session.user, "user_type") == "System User"
```
This is unused.

bhushan-barbuddhe committed:
[fix]: Hook changed to send token.

bhushan-barbuddhe committed:
chore: remove unnecessary comment

bhushan-barbuddhe committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe committed:
feat: add mobile OTP login option and set as default in system settings

- Introduced a new setting to make mobile login with OTP the default login method.
- Adjusted the login validation logic to account for the new mobile OTP option.

bhushan-barbuddhe committed:
feat: enhance mobile OTP login functionality

- Added mobile OTP login form and integrated it into the login process.
- Updated login logic to conditionally display mobile OTP or username/password forms based on system settings.

bhushan-barbuddhe committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe committed:
feat: add mobile OTP login option and set as default in system settings

- Introduced a new setting to make mobile login with OTP the default login method.
- Adjusted the login validation logic to account for the new mobile OTP option.

feat: enhance mobile OTP login functionality

- Added mobile OTP login form and integrated it into the login process.
- Updated login logic to conditionally display mobile OTP or username/password forms based on system settings.

refactor: clean up mobile OTP login implementation

- Removed deprecated mobile OTP login settings and related UI elements.
- Streamlined the login process to enhance clarity and maintainability.
- Updated the login template to conditionally render mobile OTP fields based on system settings.

bhushan-barbuddhe committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe committed:
Merge branch 'develop' of https://github.com/dhwani-ris/frappe into develop

bhushan-barbuddhe committed:
feat: enhance mobile OTP login process and validation

- Updated the login manager to support mobile OTP authentication alongside traditional username/password login.
- Refined system settings to manage mobile login options more effectively.
- Improved error handling and session management for mobile OTP logins.

bhushan-barbuddhe committed:
refactor: streamline mobile OTP login implementation and enhance validation

- Removed deprecated comments and cleaned up the mobile OTP authentication logic in the LoginManager.
- Updated the system settings validation for mobile OTP to use cached values for better performance.
- Simplified the mobile OTP sending process and improved error handling in the send_mobile_otp function.

dhwani-ankit committed:
Merge pull request #4 from dhwani-ris/feat/passwordless-mobile-login

Feat/passwordless mobile login

NagariaHussain commented:
Add screenshot for OTP input, does it have resending logic?

NagariaHussain commented on frappe/auth.py:
```diff
@@ -249,6 +251,13 @@ def clear_active_sessions(self):
+
 		if not (user and pwd):
 			user, pwd = frappe.form_dict.get("usr"), frappe.form_dict.get("pwd")
 		if not (user and pwd):
```
```suggestion
         if not (user and pwd):
		      user, pwd = frappe.form_dict.get("usr"), frappe.form_dict.get("pwd")
		otp, tmp_id = frappe.form_dict.get("otp"), frappe.form_dict.get("tmp_id")

		if (otp and tmp_id) and not (user and pwd):
			return self._authenticate_mobile_otp(otp, tmp_id)
		
		if not (user and pwd):
```

NagariaHussain commented on frappe/twofactor.py:
```diff
@@ -158,11 +158,14 @@ def confirm_otp_token(login_manager, otp=None, tmp_id=None):
 	if not otp_secret:
 		raise ExpiredLoginException(_("Login session expired, refresh page to retry"))
 
+	# Standard 2FA verification
```
Remove this

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,89 @@
+
+def validate_mobile_otp_prerequisites() -> None:
+	if not is_mobile_otp_login_enabled():
+		frappe.throw(_("Mobile OTP login is not enabled."), frappe.AuthenticationError)
```
```suggestion
		frappe.throw(_("Phone OTP login is not enabled."), frappe.AuthenticationError)
```

Should we call it Phone instead of mobile? What do you think @akhilnarang ?

akhilnarang replied:
Yep `Phone` does seem to sound _slightly_ better here

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,89 @@
+
+def generate_mobile_otp(user: str) -> tuple[int, str]:
+	otp_secret = get_otpsecret_for_(user)
+
```
```suggestion
```

NagariaHussain commented on frappe/utils/mobile_otp.py:
```diff
@@ -0,0 +1,89 @@
+	if not status:
+		frappe.throw(_("Failed to send OTP. Please try again."), frappe.AuthenticationError)
+
+	masked_mobile = mobile_no[:4] + "******" + mobile_no[-3:] if len(mobile_no) > 7 else "******"
```
If also used in 2FA, lets extract out a common utility func, also e.g. in function doc.

NagariaHussain commented on frappe/www/login.py:
```diff
@@ -216,3 +222,39 @@ def sanitize_redirect(redirect: str | None) -> str | None:
+	import re
+
+	mobile_no = re.sub(r"[^\d+]", "", mobile_no)
+
```
Let's precompile the re for performance.

bhushan-barbuddhe committed:
feat: add mobile number masking utility and integrate into OTP process

- Introduced a new utility function to mask mobile numbers for enhanced privacy.
- Updated the mobile OTP sending logic to utilize the new masking function, improving the display of mobile numbers in notifications.
- Refactored mobile number validation in the login process to ensure proper formatting before OTP generation.

bhushan-barbuddhe committed:
refactor: improve login authentication logic and error messages

dhwani-ankit committed:
Merge pull request #5 from dhwani-ris/feat/passwordless-mobile-login

Feat/passwordless mobile login

bhushan-barbuddhe committed:
refactor: update mobile number masking function and enhance SMS display

- Renamed the mobile number masking function to a more generic `mask_string` for broader use.
- Updated the SMS display styling in the login process for improved readability.

dhwani-ankit committed:
Merge pull request #6 from dhwani-ris/feat/passwordless-mobile-login

refactor: update mobile number masking function and enhance SMS display

bhushan-barbuddhe commented:
@NagariaHussain requested changes completed

dhwani-ankit committed:
Merge branch 'frappe:develop' into develop

bhushan-barbuddhe commented on frappe/utils/__init__.py:
```diff
@@ -1202,3 +1202,21 @@ def __get__(self, instance, owner=None) -> T:
 # end: custom cached_property implementation
+
+
+def mask_string(input_string: str, mask_char: str = "*", show_first: int = 4, show_last: int = 3) -> str:
```
Add doc string

NagariaHussain commented:
Fix semgrep failing due to `cache.get` usage

bhushan-barbuddhe commented:
closing this for resolve conflicts. will share seprate PR
